src/channel-qvm.lisp

```diff
@@ -0,0 +1,374 @@
+    :accessor noise-predicate
+    :initform nil 
+    :documentation "a function that is true when an instruciton is matched, false otherwise")
+   (operation-elements
+    :initarg :operation-elements
+    :accessor operation-elements
+    :initform ()  
+    :documentation "operation elements (kraus operators) to apply when the predicate is satisfied")))
+
+
+(defun make-noise-rule (predicate operation-elements)
+  "Returns a noise rule with the specified prediate and operation elements. "
+  (make-instance 'noise-rule :noise-predicate predicate :operation-elements operation-elements))
+
+
+;; A noise model consists of a set of noise rules and readout-povms. The noise rules define quantum channels and where they should be
+;; applied,  and the readout-povms describe the noisy readout probabilities. 
+(defclass noise-model ()
+  ((noise-rules
+    :initarg :noise-rules
+    :accessor noise-rules
+    :initform nil
+    :documentation "A list of noise-rules (predicate channel pairs)")
+   (readout-povms
+    :initarg :readout-povms
+    :accessor readout-povms
+    :initform (make-hash-table :test 'eql)
+    :documentation "A map of qubit to readout noise assignment probabilities. These are treated differently from noise channels (for now).")))
+
+
+(defun make-noise-model (noise-rules)
+  "Returns a noise model with noise-rules."
+  (make-instance 'noise-model :noise-rules noise-rules))
+
+
+(defmethod add-noise-models (nm1 nm2)
+  " Combines two noise models by concatenating their rules. This way, either rule can be matched by an instruction. "
+  (make-noise-model (append (noise-rules nm1) (noise-rules nm2))))
+
+
+(defmethod multiply-noise-models (nm1 nm2)
+  "Combines two noise models in a way such that both set of rules, or one and not the other could be matched by an instruction."
+  (let ((nrs (loop :for rule1 in (noise-rules nm1) 
+                   :append (loop :for rule2 in (noise-rules nm2)
+                                 :collect (make-noise-rule 
+                                           (predicate-and (noise-predicate rule1) (noise-predicate rule2)) 
+                                           (list (operation-elements rule1) (operation-elements rule2)))
+                                 :collect (make-noise-rule 
+                                           (predicate-and (noise-predicate rule1) (predicate-not (noise-predicate rule2))) 
+                                           (operation-elements rule1))
+                                 :collect (make-noise-rule 
+                                           (predicate-and (predicate-not (noise-predicate rule1)) (noise-predicate rule1)) 
+                                           (operation-elements rule2))))))
+    (make-noise-model nrs)))
+
+
+;; The following are examples of predicates! 
+(defun match-strict-qubits (qubits)
+  "The returned function is true if the instruciton's qubits are exactly equal to qubits"
+  (lambda (i-qubits i-name) (equal qubits i-qubits)))
```

Comment:
When we wrote the pyquil version of this we could be a bit sloppy because quil in pyquil has less of a clear AST. But here we really should have predicates working on `quil:instruction` objects (see class definitions in `quilc/src/ast.lisp`). For example, this would be something like

```
(lambda (instr)
  (and (typep instr 'quil:gate-application)
           (equal qubits (mapcar #'quil:qubit-index (gate-application-arguments instr)))))
```
or something like that
```

Reply:
Yeah this sounds good. It's definitely cleaner and the way I was doing it before gave me "unused argument" compilation warnings if I didn't use both lambda arguments in the lambda statement.

Reply:
Just for future reference, you can silence those unused variable warnings with a `(declare (ignore blah))`. In this instance, e.g.,

```lisp
(lambda (i-qubits i-name) 
  (declare (ignore i-name))
  (equal qubits i-qubits))
```